sync3/caches/user.go

```diff
@@ -263,6 +270,17 @@ func (c *UserCache) OnRegistered(ctx context.Context, _ int64) error {
 				})
 			}
 		}
+
+		// Record when we joined the room. We've just had to scan the history of our
+		// membership in this room to produce joinedRooms above, so we may as well
+		// do this here too.
+		c.roomToDataMu.Lock()
+		urd, ok := c.roomToData[room.RoomID]
+		if !ok {
+			urd = NewUserRoomData()
+		}
+		urd.JoinNID = joinNIDs[room.RoomID]
+		c.roomToDataMu.Unlock()
```

Comment:
Would it make more sense to have a second loop whose only job is to write `JoinNID`, and to take the lock outside of the loop?

Reply:
This is fine for now, even though it will be locking N times (once per room). We'd see it in the profiler if it's causing latency problems.